hw/xbox/xid-sdl.c

```diff
@@ -195,35 +193,21 @@ static const XIDDesc desc_xid_xbox_gamepad = {
 #define BUTTON_MASK(button) (1 << ((button) - GAMEPAD_DPAD_UP))
 
 static void update_output(USBXIDState *s) {
-#ifdef FORCE_FEEDBACK
     if (s->sdl_haptic == NULL) { return; }
+    if (s->sdl_haptic_effect == NULL) { return; }
 
-    SDL_HapticLeftRight effect = {
-        .type = SDL_HAPTIC_LEFTRIGHT,
-        .length = SDL_HAPTIC_INFINITY,
-        /* FIXME: Might be left/right inverted */
-        .large_magnitude = s->out_state.right_actuator_strength,
-        .small_magnitude = s->out_state.left_actuator_strength
-    };
+    if (s->sdl_haptic_effect->type == SDL_HAPTIC_LEFTRIGHT) {
+        s->sdl_haptic_effect->leftright.large_magnitude = s->out_state.right_actuator_strength;
+        s->sdl_haptic_effect->leftright.small_magnitude = s->out_state.left_actuator_strength;
+    }
 
-    if (s->sdl_haptic_effect_id == -1) {
-        int effect_id = SDL_HapticNewEffect(s->sdl_haptic,
-                                            (SDL_HapticEffect*)&effect);
-        if (effect_id == -1) {
-            fprintf(stderr, "Failed to upload haptic effect!\n");
-            SDL_HapticClose(s->sdl_haptic);
-            s->sdl_haptic = NULL;
-            return;
-        }
-        SDL_HapticRunEffect(s->sdl_haptic, effect_id, 1);
+    if (s->sdl_haptic_effect->type == SDL_HAPTIC_SINE) {
+        s->sdl_haptic_effect->periodic.magnitude =
+            (s->out_state.right_actuator_strength + s->out_state.left_actuator_strength) >> 1;
+    }
 
-        s->sdl_haptic_effect_id = effect_id;
 
-    } else {
-        SDL_HapticUpdateEffect(s->sdl_haptic, s->sdl_haptic_effect_id,
-                               (SDL_HapticEffect*)&effect);
-    }
-#endif
+    SDL_HapticUpdateEffect(s->sdl_haptic, s->sdl_haptic_effect_id, s->sdl_haptic_effect);
```

Comment:
With my testing, I've found just having this in the `update_output` function causes serious performance regression.

Maybe there can be improvements with only updating the effect when the actuator value changes? Or maybe run this in a non-blocking manner?

Reply:
False alarm on the regression. After further testing and capturing timing on these functions, turns out the performance drop was from other causes, and I just had bad luck when I turned force feedback off making me feel like I introduced a regression.